components/application-mgt/org.wso2.carbon.device.application.mgt.core/src/main/java/org/wso2/carbon/device/application/mgt/core/impl/SubscriptionManagerImpl.java

```diff
@@ -142,32 +100,52 @@
-                    log.error("Device with ID: " + device.getId() + " not found to install the application.");
+                DeviceManagementProviderService dmpService = HelperUtil.getDeviceManagementProviderService();
+                if (!dmpService.isEnrolled(deviceIdentifier)) {
+                    log.error("Device with ID: [" + device.getId() + "] is not enrolled to install the application.");
                 } else {
                     if (log.isDebugEnabled()) {
-                        log.debug("Prepare application install to : " + device.getId());
+                        log.debug("Installing application to : " + device.getId());
                     }
-                    activeDeviceList.add(deviceIdentifier);
-                    DAOFactory.getSubscriptionDAO().addDeviceApplicationMapping(device.getId(), applicationUUID, false);
+                    //Todo: generating one time download link for the application and put install operation to device.
+
+                    // put app install operation to the device
+                    ProfileOperation operation = new ProfileOperation();
+                    operation.setCode("INSTALL_APPLICATION");
+                    operation.setType(Operation.Type.PROFILE);
+                    operation.setPayLoad("{'type':'enterprise', 'url':'http://10.100.5.76:8000/app-debug.apk', 'app':'" + applicationUUID + "'}");
+                    List<org.wso2.carbon.device.mgt.common.DeviceIdentifier> deviceIdentifiers = new ArrayList<>();
+                    deviceIdentifiers.add(deviceIdentifier);
+                    dmpService.addOperation(DeviceManagementConstants.MobileDeviceTypes.MOBILE_DEVICE_TYPE_ANDROID,
+                            operation, deviceIdentifiers);
+
+                    DeviceApplicationMapping deviceApp = new DeviceApplicationMapping();
+                    deviceApp.setDeviceIdentifier(device.getId());
+                    deviceApp.setApplicationUUID(applicationUUID);
+                    deviceApp.setVersionName(versionName);
+                    deviceApp.setInstalled(false);
+                    dmpService.addDeviceApplicationMapping(deviceApp);
+//                    DeviceManagementDAOFactory.openConnection();
+//                    DAOFactory.getSubscriptionDAO().addDeviceApplicationMapping(device.getId(), applicationUUID, false);
                     failedDeviceList.remove(device);
                 }
-            } catch (DeviceManagementDAOException | SQLException e) {
-                throw new ApplicationManagementException("Error locating device.", e);
-            } finally {
-                DeviceManagementDAOFactory.closeConnection();
+            } catch (DeviceManagementException | OperationManagementException | InvalidDeviceException e) {
+                log.error("Error while installing application to device[" + deviceIdentifier.getId() + "]", e);
             }
+//            finally {
```

Comment:
If the connection is not opened here, and you wanted to remove it, then please remove the code block instead of commenting it.

Reply:
This is a code that is needed later on. Put it there as a reference and commented out. This will be uncommented once the DAO layers are implemented.